pkg/retriever/combinators/splitretriever.go

```diff
@@ -2,58 +2,61 @@ package combinators
 // to multiple retrievers using the given coordination style
-type SplitRetriever struct {
-	CandidateSplitter   types.CandidateSplitter
-	CandidateRetrievers []types.CandidateRetriever
-	CoordinationKind    types.CoordinationKind
+type SplitRetriever[T comparable] struct {
+	AsyncCandidateSplitter types.AsyncCandidateSplitter[T]
+	CandidateRetrievers    map[T]types.CandidateRetriever
+	CoordinationKind       types.CoordinationKind
 }
 
-func (m SplitRetriever) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetrieval {
-	candidateRetrievals := make([]types.CandidateRetrieval, 0, len(m.CandidateRetrievers))
-	for _, candidateRetriever := range m.CandidateRetrievers {
-		candidateRetrievals = append(candidateRetrievals, candidateRetriever.Retrieve(ctx, request, events))
+func (m SplitRetriever[T]) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetreival {
+	candidateRetrievals := make(map[T]types.CandidateRetreival, len(m.CandidateRetrievers))
+	for key, candidateRetriever := range m.CandidateRetrievers {
+		candidateRetrievals[key] = candidateRetriever.Retrieve(ctx, request, events)
 	}
-	return splitRetrieval{
-		retrievalSplitter:   m.CandidateSplitter.SplitRetrieval(ctx, request, events),
+	return splitRetrieval[T]{
+		retrievalSplitter:   m.AsyncCandidateSplitter.SplitRetrievalRequest(ctx, request, events),
 		candidateRetrievals: candidateRetrievals,
 		coodinationKind:     m.CoordinationKind,
 		ctx:                 ctx,
 	}
 }
 
-type splitRetrieval struct {
-	retrievalSplitter   types.RetrievalSplitter
-	candidateRetrievals []types.CandidateRetrieval
+type splitRetrieval[T comparable] struct {
+	retrievalSplitter   types.AsyncRetrievalSplitter[T]
+	candidateRetrievals map[T]types.CandidateRetreival
 	coodinationKind     types.CoordinationKind
 	ctx                 context.Context
 }
 
-func (m splitRetrieval) RetrieveFromCandidates(candidates []types.RetrievalCandidate) (*types.RetrievalStats, error) {
-	splitCandidates, err := m.retrievalSplitter.SplitCandidates(candidates)
+func (m splitRetrieval[T]) RetrieveFromAsyncCandidates(asyncCandidates types.InboundAsyncCandidates) (*types.RetrievalStats, error) {
+	asyncSplitCandidates, errChan := m.retrievalSplitter.SplitAsyncCandidates(asyncCandidates)
+	coordinator, err := coordinators.Coordinator(m.coodinationKind)
 	if err != nil {
 		return nil, err
 	}
-	retrievers := make([]types.CandidateRetrievalCall, 0, len(splitCandidates))
-	for i, candidateSet := range splitCandidates {
-		if len(candidateSet) > 0 && i < len(m.candidateRetrievals) {
-			retrievers = append(retrievers, types.CandidateRetrievalCall{
-				Candidates:         candidateSet,
-				CandidateRetrieval: m.candidateRetrievals[i],
-			})
+	stats, err := coordinator(m.ctx, func(ctx context.Context, retrievalCall func(types.RetrievalTask)) {
```

Comment:
Where does `ctx` go? shouldn't this one go down into `retrievalCall` somehow so it can be cancelled by Race when one of them wins? Otherwise wouldn't Race return the first result, try to cancel the others, but ultimately leave them running until completion?